### ISSUE TITLE
Issues with Twitter and Instagram built-in and custom rules + a feature suggestion

### ISSUE THREAD

**Issue by darkred** [tophf/mpiv#8]
Greetings

I'd like please to report a few issues I've noticed with Twitter and Instagram built-in and custom rules, 
and make a feature suggestion.

*(using Chrome and Tampermonkey/Violentmonkey)*

---

Regarding Twitter (new layout + "autoplay" enabled in https://twitter.com/settings/data) :

Here are two videos:
(GIF/mp4) https://twitter.com/natfriedman/status/1232727363626029056
(blob) https://twitter.com/github/status/1191426884828950531

Inspecting the video nodes HTML, the former is mp4, the latter is blob:
<details>
<summary>GIF</summary>

```html
<video preload="auto" playsinline="" aria-label="Embedded video" 
poster="https://pbs.twimg.com/tweet_video_thumb/ERuG9JhVUAEglNW.jpg" 
src="https://video.twimg.com/tweet_video/ERuG9JhVUAEglNW.mp4" type="video/mp4" 
style="width: 100%; height: 100%; position: absolute; background-color: black; top: 0%; left: 0%; transform: rotate(0deg) scale(1.005);"></video>
```
</details>

<details>
<summary>blob</summary>

```html
<video preload="none" playsinline="" aria-label="Embedded video" 
poster="https://pbs.twimg.com/media/EIiwEVIVAAUWOqK.jpg" 
src="blob:https://twitter.com/59eda4ea-4178-43d4-b11a-09f26acf8203" 
style="width: 100%; height: 100%; position: absolute; background-color: black; top: 0%; left: 0%; transform: rotate(0deg) scale(1.005);"></video>
```

</details>

The rule from the repository (see below) doesn't work when hovering on the mouse on each of the videos, although it seems it be still valid *(although, the rule doesn't work in the initial MPIV neither)* .
**Could the rule below be modified to work for both videos?**

```json
{"d": "twitter.com",
 "e": "video",
 "s": "var bg = node.poster, url = /http[^\\\")]+/.exec(bg)[0];if (url.indexOf('.jpg') == -1){return url.replace('pbs', 'video').replace('_thumb', '').replace('.png', '.mp4');} else {return url.replace('pbs', 'video').replace('_thumb', '').replace('.jpg', '.mp4');}"
}
```

---

Regarding Instagram:

- Hovering on multi-image posts, the first image is displayed on hovering all images 
https://www.instagram.com/p/B8eYh4UgmsN/ *(open URL in new tab)*

- Hovering on IGTV (Instagram TV) video post thumbnails
e.g. currently the 3rd post in https://www.instagram.com/instagram/
an image is displayed instead of the video.

---

My feature suggestion (based on all the above): 

to add a setting in "MPIV", something like: "Display debug info for applied rule in console", 
which, when enabled to display in console: 
- the currently applied rule (prettified, as in MPIV settings), 
- the actual values of its various properties, e.g. `d`, `e`, `q`, etc.
- whether the applied rule rule is built-in or custom

This would help in quickly finding which rule is currently applied, 
and I believe that it would assist the user in improving/fixing the relevant rule.

&nbsp;

Thank you!

**Comment by tophf** [comment]
I've fixed the multi-image one, gonna look at the video problems later. Although I'm not sure I can do anything about twitter as they've fully reworked the way they're showing the videos about a year ago and AFAIK it's no longer embeddable outside of their internal framework.

**Comment by tophf** [comment]
* Fixed the videos on instagram.
* Fixed the gif video on twtter.
* Can't fix the blob video: even duplicating this element manually in devtools results in a playback error.
* The matched rule info will be shown in console automatically when MPIV config is open (via Tampermonkey's toolbar menu).  If the UI obstructs the page you can hide it in devtools (select it in the element inspector and press <kbd>H</kbd> or simply remove the node). As for whether the rule is built-in or not I don't think this is worth maintaining as it's something that's trivial to look up. I also don't understand what difference this distinction makes.

You can manually copypaste the code from the repository or force-reinstall the script to test it. I will only bump the version after it's tested for a while.

**Comment by darkred** [comment]
First of all thanks a lot for your quick response and your work!

---

My notes using Chrome 80 *(having enabled "Disable cache" in Chrome 'Network' devtool, to make sure the video doesn't from load cache from previous tests)*
and Tampermonkey beta 4.10.6111 / Violentmonkey 2.12.7


Regarding Instagram:

- multi-page posts: 
  - if the pages in the post are all images, then it now works ok,  in both TM + VM.
  - But, if there's a video on e.g. the 2nd 'page' of an multi-page post,
e.g. in https://www.instagram.com/p/B8jmH_tAZKC/, then the video is not displayed in any case.
- IGTV videos + IG videos: *(the former have a 'tv' icon, the latter have a standard 'video' icon)*
  - (not having login to IG) ok with TM, **not working** with VM 
  - (having login to IG) **not working**, both TM and VM

---

Regarding the Twitter giv video:
I confirm it works ok (after you click it to start playing), in both TM and VM

---

Regarding the Twitter blob video: 
ok, thanks for the effort.

---

Regarding the matched rule info:

I like the way the info is displayed, as objects. Great!

I'd like to make just one suggestion: 
the matched rule info to be displayed in console **after some delay**, 
the same `after, ms` delay as in MPIV settings (500ms by default)
to avoid displaying irrelevant info as you move the mouse towards the wanted thumbnail.
 
---

Regarding:
> As for whether the rule is built-in or not I don't think this is worth maintaining as it's something that's trivial to look up. I also don't understand what difference this distinction makes.

in my humble opinion, I thought it's convenient to be able to quickly distinguish them, because, the way I see it, built-in MPIV rules are already perfect, or near perfect, because they are constantly maintained by you, the authors of this fork, who are very experienced in programming in general.

The above comes in contrast to rules from the repository, where a rule might be old, or just written by an inexperienced user, and can be improved or should be rewritten from scratch.

**Comment by tophf** [comment]
>video on e.g. the 2nd 'page' of an multi-page post

Fixed.

> IGTV videos

Works for me in VM. I don't have an account on instagram though.

> to avoid displaying irrelevant info as you move the mouse towards the wanted thumbnail

When debugging what's wrong that info is actually relevant and crucial.

> built-in MPIV rules are already perfect

Nope, I maintain only a few rules that I use myself.

**Comment by tophf** [comment]
I've slightly tweaked the debug log. Should spam less.

**Comment by darkred** [comment]
Thank you !

---

> Works for me in VM. I don't have an account on instagram though.

I found out exactly what is going on with both IGTV and IG videos (the same is for both TM and VM):
the rule only works as long as you don't login.
It doesn't work after you login, and also it might also not work after logout.

STR
- In a fresh Chrome profile, install TM or VM and the MPIV
- open https://www.instagram.com/instagram/
and firstly close the two introductory notices, one on top and the other on bottom of page
- mouseover any IGTV ![2020-02-29_172544](https://user-images.githubusercontent.com/723651/75610571-156e0d80-5b1b-11ea-94cd-e028d8de6a17.jpg)  and and IG ![2020-02-29_172530](https://user-images.githubusercontent.com/723651/75610573-1b63ee80-5b1b-11ea-95fe-ef0e94a3e9e9.jpg) video: all are displayed ok,
- now login to your account: the rule no longer works (`File not found`) for both IGTV and IG videos:
![2020-02-29_182631](https://user-images.githubusercontent.com/723651/75611197-0db16780-5b21-11ea-8d02-9cff81b82cd0.jpg)
- if you now logout, the rule will work ok,
or (if you reload the page, one or a few times) it might display static images (the fullsize image of the video thumbnail) instead of the video.

&nbsp;

I understand that you don't have an instagram account, but if someone has one, 
it would be great if he could take a look.

**Comment by tophf** [comment]
Should be fixed now.

**Comment by darkred** [comment]
I've tried it in a fresh profile, and I confirm it works in all cases! Awesome!

And, about the times where sometimes was displaying static images instead of the videos after logging out,
well, I found the following simple solution: 
simply click near the middle-top area of the profile page (I guess in order to force `focus()` inside it, 
and the rule will work fine again.

&nbsp;

Thanks so much for all your help!

**Comment by tophf** [comment]
The videos didn't start playing for an unknown reason, should be fixed now.
I've also added gallery mode popup for carousel entries.

**Comment by darkred** [comment]
> The videos didn't start playing for an unknown reason, should be fixed now.

(I've updated to latest 1.1.0, testing on fresh Chrome profile)
Unfortunately this has occurred to me again a few times, randomly (once was after switching tab focus, another after reloading page), having login or not.
For some reason, clicking on the area near the profile info always fixes it.

I don't know if it helps you, because I can't reproduce it specifically,
but it occurs consistently if you follow the above STR without closing the introductory black bars: 

- In a fresh Chrome profile, install TM or VM and MPIV
- open https://www.instagram.com/instagram/
- **don't close** the introductory black bars on top and bottom 
(i.e. the *"To help personalize content, tailor and measure ads"* on top, and 
the *"Log in to see photos and videos from friends and discover other accounts you'll love." on bottom).*
- mouseover any IGTV/IG videos, you'll get static image instead of the video. 
- clicking anywhere inside the area near the profile info (e.g. somewhere below `335m followers`) fixes it.

---
> I've also added gallery mode popup for carousel entries.

Awesome! Thanks!

Could you please make the gallery mode also work for carousel entries **when opened in a new tab**? e.g. open in new tab this: https://www.instagram.com/p/B8jmH_tAZKC/
Currently, after you mouseover the first image, if you mousewheel down or press `k` it just closes the popup.

&nbsp;

Also, I've noticed in pages like https://www.instagram.com/andyjamesguitar/ or https://www.instagram.com/baklesi_lila/, 
that hovering on any carousel post (both having login or not) results in:
`Unexpected token � in JSON at position 0`

&nbsp;

P.s. I'm reopening this.

**Comment by Ryujinra** [comment]
> The videos didn't start playing for an unknown reason, should be fixed now.

I've also experienced this, and not just on `instagram.com`. Gfycat links also do not start playing automatically on `old.reddit.com`, but I can reproduce on `instagram.com` as well for their videos.

Go here (or on an Instagram page with videos): [old.reddit.com/domain/gfycat.com](https://old.reddit.com/domain/gfycat.com/) Hover a link _without_ clicking inside the page yet - it won't autoplay. Now click inside the page and re-hover the link - it and all other links on the page will autoplay properly now. Refreshing the page will reproduce this issue every time - until you click inside the page, no video will autoplay.

I can reproduce this on `instagram.com` as well with the latest branch.

**Comment by tophf** [comment]
Videos don't start due to a restriction in the browser on autoplay: it wants you to interact with the page to confirm you're consciously present. Just moving a mouse or scrolling doesn't count. I don't think there's a way to circumvent this because it would be a huge security hole.

>carousel entries when opened in a new tab

If I understood you correctly, the answer is no - that would be wrong because it would duplicate the main functionality of the page which already has carousel controls. Historically/conceptually this is not how galleries in MPIV work.

>Unexpected token

Fixed.

**Comment by darkred** [comment]
> If I understood you correctly, the answer is no - that would be wrong because it would duplicate the main functionality of the page which already has carousel controls. Historically/conceptually this is not how galleries in MPIV work.

What I meant was: 
(Suppose you have opened in a new tab a 2-image carousel post)
As you mouseover its 1st image and that first fullsize image is shown in the popup, 
**to be able to switch in the popup to the 2nd image, via mousewheel down or via pressing `k`.**

I believe you have understood me correctly. 
So, thanks for explaining and sorry for asking.

---

But, there's currently one issue, independent from the above: 
**for carousel posts, the popup shows the 1st image/video of the post for all following images,**
independently on having opened the post as overlay or in a new tab.

STR
- open https://www.instagram.com/instagram/
- locate the first 2-image carousel post: it's 8th row, 3rd column ([image](https://user-images.githubusercontent.com/723651/75636059-222a5880-5c24-11ea-81c3-74d7e1178f23.jpg))
- Either click that thumbnail to open the carousel post as overlay on top of https://www.instagram.com/instagram/
or rightclick it and open that link in new tab: https://www.instagram.com/p/B8jmH_tAZKC/
- Mouseover on its 1st image: it will be displayed ok in the popup. 
- Click the page's "Next" button ![2020-03-02_005034](https://user-images.githubusercontent.com/723651/75635572-ef7e6100-5c1f-11ea-91b7-69625f0dcabf.jpg) to get to the 2nd image of the post and mouseover it: 
the popup will still be the 1st image.
- The same occurs for any carousel posts I've opened in new tab.

---

> Fixed.

Thanks once more.

**Comment by tophf** [comment]
>So, thanks for explaining and sorry for asking.

You can write your own rule for that. Choose an element in the post that's present on carousels, add its selector in `"e"`. The rule would look like this IIRC: `{"d":"instagram.com", "e":"foo", "follow":true, "s":"return location.href"}`

>the popup will still be the 1st image

Fixed. I broke it somewhere along the way.

**Comment by darkred** [comment]
> Fixed. I broke it somewhere along the way.

Thanks!


> You can write your own rule for that. Choose an element in the post that's present on carousels, add its selector in `"e"`. The rule would look like this IIRC: `{"d":"instagram.com", "e":"foo", "follow":true, "s":"return location.href"}`

Thanks for the suggestion. Unfortunately, it seems I'm not able to complete the rule myself : 

take for example: https://www.instagram.com/p/B8jmH_tAZKC/ *(open in new tab)*
I've tried using various selectors in "e",
all for the `<img>` element in the carousel that contains the full URL (in their "src" property), e.g. `img[decoding=\"auto\"]`,
with your rule becoming: `{"d": "instagram.com", "e": "img[decoding=\"auto\"]",  "follow": true, "s": "return location.href"}`
but unfortunately it doesn't switch to the next image in the popup.
when I mousewheel down or pres `k` while mouseover the 1st image. 

Also, when I click the page's "Next" button, there's no redirection in the address bar,
so maybe the `"follow":true`option in your rule doesn't help in this case?

Could you please help me one last time, with this one?

**Comment by tophf** [comment]
That's because the site has an overlay `div` above the image - look at the console output and you'll see it in `node`. You can adjust the selector to match some parent element or choose an entirely different element on the page as I suggested initially.

Anyway I couldn't figure out how to relay the result to the built-in rule so I can only suggest copying it:
```json
{"d": "instagram.com",
 "e": "article div",
 "s": "return location.pathname.startsWith('/p/') && (location.href + '?__a=1')",
 "g": "let media = JSON.parse(text).graphql.shortcode_media, items = media.edge_sidecar_to_children.edges.map(e => ({url: e.node.video_url || e.node.display_url || '', desc: e.node.accessibility_caption || ''})); try { items.title = media.edge_media_to_caption.edges[0].node.text } catch(e){} return items"
}
```

**Comment by darkred** [comment]
Thanks you!!
Everything works great, far better than I could hope for!

**Comment by darkred** [comment]
Hi, @tophf

I've noticed the following thing:
for instagram carousel posts, the image's `<alt>` attribute is also included in the text that's displayed in the bar on top,
e.g. "Image may contain: sky, cloud, etc" 
e.g. in the 1st such post in in https://www.instagram.com/explore/tags/landscapes/ it's:
![Screenshot (1553)](https://user-images.githubusercontent.com/723651/76451727-8db8b680-63d8-11ea-88d1-b1447e344039.jpg)

**Could this please be omitted?**

---

And, I wanted to report something different to the issue:
<sup>(I thought to post it here instead of making a separate one)</sup>

&nbsp;

it's great that you have added a progress indicator (`onprogress: Remoting.getImageProgress`)
in the `getImage()` and `saveFile()` functions.

But, I've tried mouseovering or downloading (<kbd>D</kbd>) on various images,
e.g. for the sake of testing I've tried thumbs to huge images e.g. [Google Search | Images | over 70 MP](https://www.google.com/search?as_st=y&tbm=isch&hl=en&as_q=landscape&as_epq=&as_oq=&as_eq=&cr=&as_sitesearch=&safe=images&tbs=isz:lt,islt:70mp)
but I don't see any progress counter on the black bar on top.

**Why is this happening? Does this feature not work on google search image?**
Or it's something else that I miss?

**Comment by tophf** [comment]
>Could this please be omitted

Done. I guess this text is always something useless.

>I don't see any progress

It's displayed only when downloading takes more than 3 seconds so you need a really big video or a very slow server. Images are usually cached by the browser so they're saved instantly.

**Comment by darkred** [comment]
Thank you very much!

**Comment by darkred** [comment]
> It's displayed only when downloading takes more than 3 seconds so you need a really big video or a very slow server. Images are usually cached by the browser so they're saved instantly.

Regarding the above, sorry for returning to it, but I still have some questions:

in the previous link [Google Search | Images | over 70 MP](https://www.google.com/search?as_st=y&tbm=isch&hl=en&as_q=landscape&as_epq=&as_oq=&as_eq=&cr=&as_sitesearch=&safe=images&tbs=isz:lt,islt:70mp) almost all the relevant full size images are huge files, of many MBs. They all take more than 3 sec to display or download - I can see the images get displayed progressively and take a lot of time. *This can also be confirmed from inside the Chrome's 'Network' devtool.* **So, shouldn't the progress percentage be displayed for (almost) all these images?**
I'm asking because currently the progress percentage is displayed for only a few of theses thumbnails. (I've counted 2-3 from the 24 displayed thumbnails from the page).


And, as a test:
if I keep the Chrome's 'Network' devtool open as I mouseover the thumbnails from the above link, and:
- tick the `Disable cache` tickbox and/or 
- set the throttling dropdown to the slowest available preset, "Slow 3G" preset, or create an even slower one,

**shouldn't the progress percentage be displayed for all images, in this case?**

**Comment by tophf** [comment]
It means those pictures are downloaded in big chunks so the progress isn't reported by the browser during the process. It probably depends on the server.

**Comment by darkred** [comment]
Thanks again 👍

**Comment by darkred** [comment]
I've noticed the following when browsing private instagram profiles:

the activated rule is https://github.com/tophf/mpiv/blob/master/script.user.js#L1217
This rule appends a `?__a=1` to the thumbnail URL, 
e.g. `https://www.instagram.com/p/_________________/?__a=1`
and a JSON is retrieved from that URL.
The issue is that this URL can only be loaaded in the browser when you have logged in, otherwise (or if you try to load it in an ignognito window) it gives a 404 (`Server error 404 undefined` in Console).
Could it be possible to make it work?

PS. I can't give you an actual private instagram account URL, because, you have told me that you don't have an instagram accoumt, but even if you had, there's no point in giving you a URL of a private profile that I can view, but you can't have access to.

**Comment by tophf** [comment]
I don't see any way to make it work. It seems conceptually impossible to me.

**Comment by darkred** [comment]
I found that changing line 1247 to `anonymous: true` it fixes the issue.
Would it be ok for you if we changed it to the script? I could gladly make a PR if you prefer.

**Comment by tophf** [comment]
Does it fix the issue when browsing in normal profile as well?

**Comment by tophf** [comment]
See if removing the entire `anonymous: true` line works for you in incognito.

**Comment by darkred** [comment]
Yes, it works fine in incognito, too.

**Comment by tophf** [comment]
9ea318a1
Thanks.

**Comment by darkred** [comment]
Hi, @tophf 
I wanted to report that this commit
[autodetect page CSP and use xhr mode if needed](https://github.com/tophf/mpiv/commit/1a799f5c6f63ef1de921e0e3c1eb5bfe45b120e6)
has broken the displaying of:
- IG video posts  (single or multi)
- IGTV (Instagram TV) video posts

STR
- In a fresh Chrome 81 profile, install latest TM beta / VM and the MPIV of the above commit
https://github.com/tophf/mpiv/raw/1a799f5c6f63ef1de921e0e3c1eb5bfe45b120e6/script.user.js
- go to https://www.instagram.com/instagram/
- hover to: 
  - any IG video posts, single or multi(e.g. currently the [2nd post](https://user-images.githubusercontent.com/723651/81348057-09c43780-90c6-11ea-90cf-f7af9b49108c.jpg), and
  - any IGTV (Instagram TV) video post
- there won't be any video in the popup in any of these cases.
In the case of IG video posts, you'll get `File can't be displayed.` in console.

**Comment by tophf** [comment]
It's not related to CSP detection because I see the same problem in much older versions (like 1.1.7) of MPIV. 

Even if I create the `<video>` element manually in devtools inside the instagram page and specify the correct video src, it won't play.

Looking at `chrome://media-internals` I see "audio decoder initialization failed" error so Chrome thinks there's a problem in the audio config of the video but the video plays fine when opened in a new tab...

**Comment by tophf** [comment]
There was a bug in CSP detection though which I fixed.

**Comment by tophf** [comment]
After restarting the browser and using the master branch I don't see the media decoder error so I guess that error was caused by CSP detection bug after all e.g. the bug broke something in Chrome's internal media environment.

**Comment by darkred** [comment]
I don't see the issue anymore after your new commit. Thank you.

**Comment by darkred** [comment]
@tophf 
Today I've noticed that the gallery mode popup for carousel entries no longer works,
e.g. in https://www.instagram.com/facebook/ and hover over [the newest post](https://user-images.githubusercontent.com/723651/113734384-f8cfdc00-9703-11eb-8881-37a691fafb1e.jpg) :
there is no counter in the [title](https://user-images.githubusercontent.com/723651/113734468-08e7bb80-9704-11eb-9d67-553f910c022d.jpg) ( it should be `[1/3]` )

I've tested in a fresh Chrome/Firefox profile with TM/VM.
There's no error in console.
Occurs both login or not.
Tried various previous script versions and it occurs in all, it's not a regression .
